apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java

```diff
@@ -128,52 +167,169 @@ public T getForClassLoaderOfClass(Class<?> classOfTargetClassLoader) {
             return localHelper;
         }
 
-        private static <T> T createHelper(@Nullable ClassLoader targetClassLoader, ElasticApmTracer tracer, String implementation, String... additionalHelpers) {
-            try {
-                final Map<String, byte[]> typeDefinitions = getTypeDefinitions(asList(implementation, additionalHelpers));
-                Class<? extends T> helperClass;
-                try {
-                    helperClass = loadHelperClass(targetClassLoader, implementation, typeDefinitions);
-                } catch (ClassNotFoundException | NoClassDefFoundError e) {
-                    // in the unit tests, the agent is not added to the bootstrap class loader
-                    helperClass = loadHelperClass(ClassLoader.getSystemClassLoader(), implementation, typeDefinitions);
+        @Override
+        protected void markFailedClassLoader(ClassLoader failedClassLoader) {
+            failed = true;
+        }
+    }
+
+    public static class ForAnyClassLoader<T> extends HelperClassManager<T> {
+
+        // doesn't need to be concurrent - invoked only from a synchronized context
+        static final Map<Integer, WeakReference<List<Object>>> clId2helperImplListMap = new HashMap<>();
+
+        final ConcurrentHashMap<Integer, WeakReference<T>> clId2helperMap;
+        final ConcurrentHashMap<Integer, Boolean> failedClassLoaderSet;
```

Comment:
Have you considered `Collections.newSetFromMap` or `WeakConcurrentSet` here?

Reply:
No. I assume `WeakConcurrentSet` has no added value if I reuse the boxed CL hash code.
As for `Collections.newSetFromMap` - good idea, seems to be no added overhead and with the set API

Reply:
No need for it anymore now that I use weak keys for the class loader in the other map- I can just map the CL to a `WeakReference` wrapping null for caching